crates/ra_syntax/src/syntax_error.rs

```diff
@@ -1,209 +1,47 @@
-//! FIXME: write short doc here
+//! Module that defines `SyntaxError`.
 
 use std::fmt;
 
-use ra_parser::ParseError;
-
-use crate::{validation::EscapeError, TextRange, TextUnit, TokenizeError};
+use crate::{TextRange, TextUnit};
 
+/// Represents the result of unsuccessful tokenization, parsing
+/// or semmantical analyzis.
 #[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct SyntaxError {
-    kind: SyntaxErrorKind,
-    location: Location,
-}
-
-// FIXME: Location should be just `Location(TextRange)`
-// TextUnit enum member just unnecessarily compicates things,
-// we should'n treat it specially, it just as a `TextRange { start: x, end: x + 1 }`
-// see `location_to_range()` in ra_ide/src/diagnostics
-#[derive(Clone, PartialEq, Eq, Hash)]
-pub enum Location {
-    Offset(TextUnit),
-    Range(TextRange),
-}
-
-impl From<TextUnit> for Location {
-    fn from(offset: TextUnit) -> Location {
-        Location::Offset(offset)
-    }
-}
-
-impl From<TextRange> for Location {
-    fn from(range: TextRange) -> Location {
-        Location::Range(range)
-    }
-}
-
-impl fmt::Debug for Location {
-    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
-        match self {
-            Location::Offset(it) => fmt::Debug::fmt(it, f),
-            Location::Range(it) => fmt::Debug::fmt(it, f),
-        }
-    }
-}
+pub struct SyntaxError(String, TextRange);
+
+// FIXME: there was an unused SyntaxErrorKind previously (before this enum was removed)
+// It was introduced in this PR: https://github.com/rust-analyzer/rust-analyzer/pull/846/files#diff-827da9b03b8f9faa1bade5cdd44d5dafR95
+// but it was not removed by a mistake.
+//
+// So, we need to find a place where to stick validation for attributes in match clauses.
+// Code before refactor:
+// InvalidMatchInnerAttr => {
+//    write!(f, "Inner attributes are only allowed directly after the opening brace of the match expression")
+// }
 
 impl SyntaxError {
-    pub fn new<L: Into<Location>>(kind: SyntaxErrorKind, loc: L) -> SyntaxError {
-        SyntaxError { kind, location: loc.into() }
+    pub fn new(message: impl Into<String>, range: TextRange) -> Self {
+        Self(message.into(), range)
     }
-
-    pub fn kind(&self) -> SyntaxErrorKind {
-        self.kind.clone()
+    pub fn new_at_offset(message: impl Into<String>, offset: TextUnit) -> Self {
+        Self(message.into(), TextRange::offset_len(offset, 1.into()))
     }
 
-    pub fn location(&self) -> Location {
-        self.location.clone()
+    pub fn message(&self) -> &str {
```

Comment:
We should remove this function and just `impl fmt::Display`

Reply:
Hmm, yes, seems reasonable.

Reply:
I would expect a `fmt::Display` impl to give both the error message and location, while this method only gives the message, not the location.

Reply:
You never see error location displayed as a hint on top of error squiggles, but you do want to see the text range if you print it to stdout. In case of `std::fmt::Display` this is for the first case, whilst `std::fmt::Debug` is for the second case. But anyway, I think this is not transparent, should we remove `impl Display` at all here then?

Reply:
I did remove `message()` function, though this is a bit controversial, but I'd like to trust matklad here. Maybe I am wrong saying `not transparent` in my previous comment right here...